File: 백준/Gold/1202. 보석 도둑/보석 도둑.java
import java.lang.reflect.Array;
import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
	static BufferedReader br;
	static BufferedWriter bw;
	static StringTokenizer st;

	public static void main(String[] args) throws Exception {

		br = new BufferedReader(new InputStreamReader(System.in));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);     // 보석의 개수
        int k = Integer.parseInt(input[1]);     // 가방의 개수

        Jewelry[] jewelries = new Jewelry[n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int weight = Integer.parseInt(st.nextToken());
            int value = Integer.parseInt(st.nextToken());
            jewelries[i] = new Jewelry(weight, value);
        }

        int[] bags = new int[k];
        for (int i = 0; i < k; i++) {
            bags[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(jewelries, (o1, o2) -> {
            return o1.weight - o2.weight;
        });
        Arrays.sort(bags);

        PriorityQueue<Jewelry> candidate = new PriorityQueue<>((o1, o2) -> {
            return o2.value - o1.value;
        });
        int jIdx = 0;
        long answer = 0L;
        for (int w : bags) {
            while (jIdx < n && jewelries[jIdx].weight <= w) candidate.add(jewelries[jIdx++]);
            if (candidate.isEmpty()) continue;
            int value = candidate.poll().value;
            answer += value;
        }

        bw.write(String.valueOf(answer));
		
		bw.flush();
		bw.close();
		br.close();
	}
}

class Jewelry {
    int weight;
    int value;

    Jewelry (int weight, int value) {
        this.weight = weight;
        this.value = value;
    }
}

File: 백준/Gold/1202. 보석 도둑/보석 도둑.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;

class Jewelry {
    public:
        int weight;
        int value;

        Jewelry() : weight(0), value(0) {}
        Jewelry(int w, int v) : weight(w), value(v) {}
};

struct CompareWeight {
    bool operator() (const Jewelry& j1, const Jewelry& j2) {
        return j1.weight < j2.weight;
    }
};

struct CompareValue {
    bool operator() (const Jewelry& j1, const Jewelry& j2) {
        return j1.value < j2.value;
    }
};

int main() {

    int n, k;
    cin >> n >> k;

    vector<Jewelry> jewelries(n);
    for (int i = 0; i < n; i++) {
        cin >> jewelries[i].weight >> jewelries[i].value;
    }

    vector<int> bags(k);
    for (int i = 0; i < k; i++) {
        cin >> bags[i];
    }

    sort(jewelries.begin(), jewelries.end(), CompareWeight());
    sort(bags.begin(), bags.end());

    priority_queue<Jewelry, vector<Jewelry>, CompareValue> candidate;
    int idx = 0;
    long answer = 0;
    for (int b : bags) {
        while (idx < n && b >= jewelries[idx].weight) candidate.push(jewelries[idx++]);
        
        if (candidate.empty()) continue;
        answer += candidate.top().value;
        candidate.pop();
    }

    cout << answer << endl;
}